at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
I am not sure if `angular.extend` will mix prototype members into $delegate.
I think that it should actually be:

```typescript
angular.extend($delegate, targetClass.prototype, new targetClass());
```

and maybe pass `$delegate` to decorator's constructor:

```typescript
angular.extend(…, new targetClass($delegate));
```

so it can be referenced in decorating class.